### ISSUE TITLE
Creating context in scope hangs program when scope ends

### ISSUE THREAD

**Issue by SuperFluffy** [erickt/rust-zmq#92]
The following minimal example hangs and blocks execution of the program; the message “Socket returned” is never printed. I suspect this is because the context gets destroyed at the end of the scope, while the socket built on top of that context still lives on. I guess doing this is a very bad idea; the fix consists in not having the context destroyed (as can be seeen by uncommenting/commenting the lines I have indicated). It would be great if this would throw some serious error message.

``` rust
extern crate zmq;

fn main() {
    let address = Some("tcp://127.0.0.1:5556".to_string());
    // let mut context; // Uncomment to make it work;

    let mut publisher = match address.as_ref() {
        Some(address) => {
            println!("Setting up socket");

            let mut context = zmq::Context::new(); // Comment to make it work;
            // context = zmq::Context::new(); // Uncomment to make it work

            let mut publisher = context.socket(zmq::PUB).unwrap();
            assert!(publisher.bind(address).is_ok());

            println!("Returning socket");

            Some(publisher)
        }
        None => None
    };

    println!("Socket returned");

    let data: u8 = 1;

    loop {
        let s = format!("data {}", data);
        println!("Sending {}", s);
        publisher.as_mut().unwrap().send_str(&s, 0).unwrap();
    }
}
```

**Comment by petehayes102** [comment]
There was a PR quite a while ago that tied a socket's lifetime to the context: https://github.com/erickt/rust-zmq/pull/65. In this case, your code would not compile as the publisher outlives its context.

If this is a deal breaker, I'd suggest forking and patching your fork with jdpage's PR.

**Comment by rotty** [comment]
There's also #96, in which @tcosprojects suggests a solution involving using Arc<> inside the zmq bindings itself, which seems to be the most general/practical solution to me.

**Comment by rotty** [comment]
I've now merged #96, which makes this example code both compile and run.
